--- a/arch/arm/boot/dts/am335x-boneblack.dts
+++ b/arch/arm/boot/dts/am335x-boneblack.dts
@@ -7,7 +7,7 @@
 #include "am33xx.dtsi"
 #include "am335x-bone-common.dtsi"
 #include "am335x-boneblack-common.dtsi"
-#include "am335x-boneblack-hdmi.dtsi"
+#include "am335x-boneblack-gfz.dtsi"
 
 / {
 	model = "TI AM335x BeagleBone Black";
--- /dev/null
+++ b/arch/arm/boot/dts/am335x-boneblack-gfz.dtsi
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2012 Texas Instruments Incorporated - https://www.ti.com/
+ */
+
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/gpio/gpio.h>
+
+&am33xx_pinmux {
+	uart1_pins: pinmux_uart1_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x980, PIN_INPUT, MUX_MODE0)	/* P9_26 uart1_rxd */
+			AM33XX_PADCONF(0x984, PIN_OUTPUT, MUX_MODE0)	/* P9_24 uart1_txd */
+		>;
+	};
+
+	uart2_pins: pinmux_uart2_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x950, PIN_INPUT, MUX_MODE1)	/* P9_22 uart2_rxd */
+			AM33XX_PADCONF(0x954, PIN_OUTPUT, MUX_MODE1)	/* P9_21 uart2_txd */
+		>;
+	};
+
+	uart3_pins: pinmux_uart3_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x964, PIN_OUTPUT, MUX_MODE1)	/* P9_42 uart3_txd */
+		>;
+	};
+
+	uart4_pins: pinmux_uart4_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x870, PIN_INPUT, MUX_MODE6)	/* P9_11 uart4_rxd */
+			AM33XX_PADCONF(0x874, PIN_OUTPUT, MUX_MODE6)	/* P9_13 uart4_txd */
+			AM33XX_PADCONF(0x8d0, PIN_INPUT, MUX_MODE6)	/* P8_35 uart4_cts */
+			AM33XX_PADCONF(0x8d4, PIN_OUTPUT, MUX_MODE6)	/* P8_33 uart4_rts */
+		>;
+	};
+
+	uart5_pins: pinmux_uart5_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x8c4, PIN_INPUT, MUX_MODE4)	/* P8_38 uart5_rxd */
+			AM33XX_PADCONF(0x8c0, PIN_OUTPUT, MUX_MODE4)	/* P8_37 uart5_txd */
+			AM33XX_PADCONF(0x8d8, PIN_INPUT, MUX_MODE6)	/* P8_31 uart5_cts */
+			AM33XX_PADCONF(0x8dc, PIN_OUTPUT, MUX_MODE6)	/* P8_32 uart5_rts */
+		>;
+	};
+
+	spi1_pins: pinmux_spi1_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x990, PIN_INPUT, MUX_MODE3)	/* P9_31 spi1_sclk */
+			AM33XX_PADCONF(0x994, PIN_INPUT, MUX_MODE3)	/* P9_29 spi1_d0   */
+			AM33XX_PADCONF(0x998, PIN_OUTPUT, MUX_MODE3)	/* P9_30 spi1_d1   */
+			AM33XX_PADCONF(0x99c, PIN_OUTPUT, MUX_MODE3)	/* P9_28 spi1_cs0  */
+		>;
+	};
+
+	i2c1_pins: pinmux_i2c1_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x958, PIN_INPUT_PULLUP, MUX_MODE2)	/* P9_18 i2c1_sda  */
+			AM33XX_PADCONF(0x95c, PIN_INPUT_PULLUP, MUX_MODE2)	/* P9_17 i2c1_scl  */
+		>;
+	};
+
+	adxl355_pins: pinmux_adxl355_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x9a4, PIN_INPUT_PULLDOWN, MUX_MODE7)	/* P9_27 gpio3[19] */
+		>;
+	};
+
+	pps_pins: pinmux_pps_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x82c, PIN_INPUT_PULLDOWN, MUX_MODE7)	/* P8_17 gpio0[27] */
+		>;
+	};
+
+	w1_pins: pinmux_w1_pins {
+		pinctrl-single,pins = <
+			AM33XX_PADCONF(0x8b0, PIN_INPUT_PULLUP, MUX_MODE7)	/* P8_41 gpio2[10] */
+		>;
+	};
+};
+
+&uart1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart1_pins>;
+	status = "okay";
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pins>;
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_pins>;
+	status = "okay";
+};
+
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart4_pins>;
+	status = "okay";
+};
+
+&uart5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart5_pins>;
+	status = "okay";
+};
+
+&spi1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi1_pins>;
+	status = "okay";
+
+	adxl355: adxl355@0 {
+		compatible = "adi,adxl355";
+		pinctrl-names = "default";
+		pinctrl-0 = <&adxl355_pins>;
+		reg = <0>;
+		spi-max-frequency = <1000000>;
+		interrupt-parent = <&gpio3>;
+		interrupts = <19 IRQ_TYPE_EDGE_RISING>;
+		interrupt-names = "DRDY";
+	};
+};
+
+&i2c1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c1_pins>;
+	status = "okay";
+
+	dps310: dps310@77 {
+		compatible = "infineon,dps310";
+		reg = <0x77>;
+	};
+};
+
+&tscadc {
+	status = "okay";
+};
+
+&am335x_adc {
+	ti,adc-channels = <0 1 2 3 4 5 6 7>;
+};
+
+/ {
+	pps {
+		compatible = "pps-gpio";
+		pinctrl-names = "default";
+		pinctrl-0 = <&pps_pins>;
+		gpios = <&gpio0 27 GPIO_ACTIVE_HIGH>;
+		status = "okay";
+	};
+
+	onewire {
+		compatible = "w1-gpio";
+		pinctrl-names = "default";
+		pinctrl-0 = <&w1_pins>;
+		gpios = <&gpio2 10 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>;
+		status = "okay";
+	};
+};
